sql/tests: TestCreateRandomSchema failed [column "crdb_internal_idx_expr" does not exist]

sql/tests.TestCreateRandomSchema [failed](https://tanzanite.cluster.engflow.com/invocations/default/d30c8125-4a48-445d-b587-0d955986d5c4?testReportRun=1&testReportShard=5&testReportAttempt=1#targets-Ly9wa2cvc3FsL3Rlc3RzOnRlc3RzX3Rlc3Q=) on master @ [9d3dd0530818b9c5c0f7dac04e8a8de4acf1bba4](https://github.com/cockroachdb/cockroach/commits/9d3dd0530818b9c5c0f7dac04e8a8de4acf1bba4):


```
=== RUN   TestCreateRandomSchema
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestCreateRandomSchema1060254334
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    random_schema_test.go:70: CREATE TABLE table_60 ("c%pol60_0" GEOMETRY NOT NULL, "col60 _1" REGCLASS NOT NULL, col6🙃0_2 REGCLASS, "col6 0_3" BIT(28) NOT NULL, col60_4 INT4 NOT NULL, "c̸%p😪o*l60_5" STRING AS (CASE WHEN col6🙃0_2 IS NULL THEN e'\x017B\x1a\x17-oP\x19' ELSE 'k?^(' END) STORED, c̾ol60_6 STRING NOT NULL AS (lower(CAST("col6 0_3" AS STRING))) VIRTUAL, PRIMARY KEY ("col60 _1", col60_4 ASC), UNIQUE (c̾ol60_6) STORING ("c%pol60_0"), INDEX ((col60_4 + -1361448576) ASC) STORING ("c%pol60_0", "c̸%p😪o*l60_5") NOT VISIBLE, UNIQUE (c̾ol60_6 DESC, (col60_4 + -1044805803), col6🙃0_2 DESC) PARTITION BY LIST (c̾ol60_6) (PARTITION "\\xbatable_60_part_0" VALUES IN ((e't\x17',), ('',), (e'6G\x06Khb',), ('I7',), (e'K\x02Enc(AST',), (',',)), PARTITION table_6͖0_part_1 VALUES IN ((e'\x1dXx #\x7f\rg\x03',), (e'f\x16',), (e'y\x02z_\b\x079',), ('qW',), (e'?3\n',))), INDEX (lower(CAST("col6 0_3" AS STRING)) DESC), INVERTED INDEX ("c%pol60_0" ASC), UNIQUE ("col6 0_3" ASC), INDEX (col60_4 DESC) STORING ("c%pol60_0", col6🙃0_2), INVERTED INDEX ("c̸%p😪o*l60_5" gin_trgm_ops)) pq: column "crdb_internal_idx_expr" does not exist
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestCreateRandomSchema1060254334
--- FAIL: TestCreateRandomSchema (9.12s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=5</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCreateRandomSchema.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39530

--- a/pkg/sql/alter_table.go
+++ b/pkg/sql/alter_table.go
@@ -268,11 +268,17 @@ func (n *alterTableNode) startExec(params runParams) error {
 					return err
 				}
 
+				// We are going to modify the AST to replace any index expressions with
+				// virtual columns. If the txn ends up retrying, then this change is not
+				// syntactically valid, since the virtual column is only added in the descriptor
+				// and not in the AST.
+				columns := make(tree.IndexElemList, len(d.Columns))
+				copy(columns, d.Columns)
 				if err := replaceExpressionElemsWithVirtualCols(
 					params.ctx,
 					n.tableDesc,
 					tableName,
-					d.Columns,
+					columns,
 					false, /* isInverted */
 					false, /* isNewTable */
 					params.p.SemaCtx(),
@@ -282,7 +288,7 @@ func (n *alterTableNode) startExec(params runParams) error {
 				}
 
 				// Check if the columns exist on the table.
-				for _, column := range d.Columns {
+				for _, column := range columns {
 					if column.Expr != nil {
 						return pgerror.New(
 							pgcode.InvalidTableDefinition,
@@ -303,7 +309,7 @@ func (n *alterTableNode) startExec(params runParams) error {
 					StoreColumnNames: d.Storing.ToStrings(),
 					CreatedAtNanos:   params.EvalContext().GetTxnTimestamp(time.Microsecond).UnixNano(),
 				}
-				if err := idx.FillColumns(d.Columns); err != nil {
+				if err := idx.FillColumns(columns); err != nil {
 					return err
 				}
 

--- a/pkg/sql/create_table.go
+++ b/pkg/sql/create_table.go
@@ -1773,6 +1773,17 @@ func NewTableDesc(
 		return newColumns, nil
 	}
 
+	// Copies the index elements, and returns a closure to restore them back,
+	// so that any mutation to the AST is undone once this statement completes.
+	copyIndexElemListAndRestore := func(existingList *tree.IndexElemList) func() {
+		newList := make(tree.IndexElemList, len(*existingList))
+		copy(newList, *existingList)
+		restoreList := *existingList
+		*existingList = newList
+		return func() {
+			*existingList = restoreList
+		}
+	}
 	// Now that we have all the other columns set up, we can validate
 	// any computed columns.
 	for _, def := range n.Defs {
@@ -1811,6 +1822,11 @@ func NewTableDesc(
 			if err := validateColumnsAreAccessible(&desc, d.Columns); err != nil {
 				return nil, err
 			}
+			// We are going to modify the AST to replace any index expressions with
+			// virtual columns. If the txn ends up retrying, then this change is not
+			// syntactically valid, since the virtual column is only added in the descriptor
+			// and not in the AST.
+			defer copyIndexElemListAndRestore(&d.Columns)()
 			if err := replaceExpressionElemsWithVirtualCols(
 				ctx,
 				&desc,
@@ -1926,6 +1942,11 @@ func NewTableDesc(
 			if err := validateColumnsAreAccessible(&desc, d.Columns); err != nil {
 				return nil, err
 			}
+			// We are going to modify the AST to replace any index expressions with
+			// virtual columns. If the txn ends up retrying, then this change is not
+			// syntactically valid, since the virtual descriptor is only added in the descriptor
+			// and not in the AST.
+			defer copyIndexElemListAndRestore(&d.Columns)()
 			if err := replaceExpressionElemsWithVirtualCols(
 				ctx,
 				&desc,

--- a/pkg/sql/logictest/testdata/logic_test/create_table
+++ b/pkg/sql/logictest/testdata/logic_test/create_table
@@ -1068,3 +1068,22 @@ CREATE TABLE public.t_115352 (
 )
 
 subtest end
+
+
+statement ok
+
+subtest 125619
+
+statement ok
+SET inject_retry_errors_enabled=true
+
+statement ok
+CREATE TABLE t_125619 (i INT8, j INT8, INDEX ((i + j)));
+
+statement ok
+ALTER TABLE t_125619 ADD CONSTRAINT uni UNIQUE ((i + j + i))
+
+statement ok
+SET inject_retry_errors_enabled=false
+
+subtest end

